### ISSUE TITLE
Problem with lookup in the instance tree

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#2263]
**Reported by pavel&#46;grozman on 24 Jul 2018 14:03 UTC**
Example:

```
model A
  model B
    constant Integer n=1;
  equation
    assert(n==B.n, "n is not equal to B.n");
  end B;
end A;

class C = A.B(n=3);
```

When flattening class C, the reference B.n should be looked up in the instance of C. This instance does not contain element named B, so the lookup should continue in _the instance scope corresponding to the lexically enclosing class_ (as explained in **5.3.1 Simple Name Lookup**).\\
But the lexically enclosed class A needs not to be instantiated when flattening C, so there is no instance scope that corresponds to this class.\\
The class A may be temporary instantiated in order to find class B. In this case the value of B.n will be 1 and not equal to n.\\
The result is different from what was defined in version 3.3. There the lookup was defined in the tree of class definitions, so B.n was always equal to n.\\
If this was the intention, it should be mentioned in the list of non-backward-compatible changes.\\
I would propose to return to the old definition of lookup and add sentence that in the result of lookup (that is an element in the class tree) is treated as the corresponding element in the instance tree (with same meaning of "correspondence" that is assumed but not defined in version 3.4)

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/2263

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 6 Aug 2018 12:35 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 6 Aug 2018 13:22 UTC**
I can't see how n would be equal to B.n for the assert in C for any reasonable definition of the lookup rules. C has an n that is 3, but it does not contain a B, so how could B.n ever resolve to the n that is 3? Or did you mean that n would be 1 for the assert? That would mean that modifications on short class declarations had no effect. I could see a formulation that would give you a lookup error, but I'm pretty sure that's not what we had before.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 7 Aug 2018 14:51 UTC**
What was defined in 3.3 was ambiguous. B.n in the assertion could have been evaluated to 1 or to 3 by different tools and both answers could be justified by some parts of the Specification. See #1397 for the full discussion. In 3.4 the rules for Flattening were overhauled to clarify cases like this. Now B.n would be evaluated to 1 and the assertion would regardless of the modification on n in C. (Lexically enclosing class A _needs_ to be instantiated according to 5.6.1 and 5.6.2.)

I guess, technically, it is breaking backward compatibility to clarify cases that were ambiguous and were treated differently by different tools. At least for the tools whose interpretation was different from the one now accepted. Shouldn't it be handled by the tools then for which the interpretation changed?

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#x2e;grozman on 10 Aug 2018 12:54 UTC**
I can agree that the specification in 3.3 was ambiguous because it did not define when a node in the class tree should be mapped to the instance tree.

My interpretation was that it is done after lookup. As I see now, most common interpretation is the mapping is done after the lookup of the first identifier of a composite name.

But the current definition with lookup in the instance tree is also problematic.

With this definition we will skip some lexical scopes and therefore may find a definition in an outer scope that otherwise would be shadowed. Example:


```
record R
    Real x,y;
end R;

package B
   type R = Real[3];
   model M
      R x;
   end M;
end B;

B.M m;
```

Here the lookup in the instance m does not see the definition of B.R because it is missing in the instance tree. The global definition of R will be found instead. This will change the model m completely. Such a possibility makes the model checking meaningless.

I think the following proposal is more adequate:
 - The lookup of the first name is done in the class tree
 - The result is mapped to the instance tree. If the corresponding elements is missing in the instance tree there two options: either threat this as an error or process in the same way as in case of global lookup
 - The names after dot are looked in the instance node but only among the elements that correspond to elements of the corresponding class node.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 13 Aug 2018 14:37 UTC**
I can see what you mean. The text in Section 5.3.1 says:
" This lookup in each instance scope is performed as follows
 * Among declared named elements (class-definition and component-declaration) of the class (including elements inherited from base-classes)."

And the instance in your example is m, so its scope would be where it is defined. However, the instance scope is defined differently. The link to the instance scope is pointing to section 5.6.1.1. which gives the definition of the class tree and does not shed the light on the instance scope. I think it should be pointing to the Instance Tree (Section 5.6.1.2). There is a definition there:
"A node in the instance tree is the instance scope for the modifiers and elements syntactically defined in the class it is instantiated from."

And from the second paragraph in the Instance Tree section (5.6.1.2):
 "The instance tree has the following properties:
 - It contains the instantiated elements of the class definitions, with redeclarations taken into account and merged modifications applied.
 - Each instance knows its source class definition from the class tree and its modification environment.
 - Each modification knows its instance scope."
I think maybe the second item should say "Each instance tree node" to be more clear? It is not very clear what is meant by instance in this case.

In other words, the instance scope is not where the instance m is defined. The instance scope for "R x;" in the instance m is where "R x" is that node of the instance tree.

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#46;grozman on 14 Aug 2018 22:03 UTC**
Sorry, I do not understand what is the correct flattening of m according to your explanation.

Variant 1:

```
Real m.x.x;
Real m.x.y;
```

Variant 2:
```
Real[3] m.x
```

My opinion is that the lookup in the instance tree (as defined in 5.3.1) gives variant 1, but the old lookup in the class tree gives variant 2.

I don't like the variant 1 because it does not correspond to what is defined (and checked) in package B.

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 15 Aug 2018 08:36 UTC**
I cannot follow the discussion. What is the exact problem?

The semantics of the example you are discussing should be straight. Just expand it to make it simulate, i.e., assume you add equations for `B.M`:
```
package P
  record R
    Real x,y;
  end R;

  package B
     type R = Real[3];
     model M
        R x = {1,2,3};
     end M;
  end B;

  model MM
    B.M m;
  end MM;
end P;
```
The meaning must be that `x` of `P.B.M` is an array. The example is working like that in Dymola since at least version 2014 FD01.

Also the original example is straight:
```
package P
  model A
    model B
      constant Integer n=1;
    equation 
      assert(n==B.n, "n is not equal to B.n");
    end B;
  end A;

  class C = A.B(n=3);
end P;
```
Simulating it in Dymola (again since version 2014 FD01) will immediately result in the assertion failing. And that is right so. References are **always** looked up in lexical scope and in the given scope `B.n` is a constant with value 1. You can try to drop the `constant` modifier to understand the real trouble here: mixing type and instance names within a single "full qualified name" is not permitted. Either you want to instanciate (declaration/type-reference) in which case a type must be given (i.e., the full qualified name for the type of the component must only consist of type-names), or you want to reference a value within an expression (expression/value-reference) in which case a component reference must be given (i.e., the full qualified name for referencing a value must only consist of component-names). The full qualified name `B.n` without `n` being a constant is not a valid value-reference in the scope of `P.A.B`. Thus, dropping `constant` is not permitted for `P.A.B.n` due to the `B.n` value-reference within the assertion of `P.A.B`. And the reason is that its `B.n` lookup is according to **lexical scoping** which must in case of **value-references follow components/instances only**; and `P.A.B.n` is only a value-reference because it is a constant.

In other words, the whole case boils-down to a misunderstanding of permitted types of references (type ones for declarations and value ones for expressions) which is kind of special for constants. But I cannot see the case for problems with lookups in the instance tree.

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#46;grozman on 15 Aug 2018 09:28 UTC**
The problem with "B.M m" is that your rule "References are **always** looked up in lexical scope" is replaced in the version 3.4 of Modelica specification with a different rule (5.3.1):

  ... any simple name ... is looked up sequentially in each member of the ordered set **of instance scopes ... corresponding to lexically enclosing classes** until a match is found or an enclosing class is encapsulated. 

According to this rule, the type R should be searched
 - First, in instance B.M m; it does tot contain definition of R;
 - Next, in the instance scope corresponding to the lexical parent of B.M (that is class B). But the instance tree of B.M m does not contain corresponding node. The specification does not explain what to do in this case. The "naïve" interpretation is that the definition of R cannot be found (because there is no scope to search in), so the lookup should continue in the next scope;
 - Next, in the instance scope corresponding to the lexical parent of B. Here we can find the definition of R as a record.

My proposal is to return back the rule "References are **always** looked up in lexical scope" and explain how the result of lookup is mapped to the instance tree.

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 15 Aug 2018 10:03 UTC**
`B.n` is not a simple name. It is a composite name for which section 5.3.2 applies. According to 5.3.2, `B` is recognized as a class (i.e., type), for which reason the search for the rest of the name proceeds within its flattened, **unmodified** model (modifications are therefore not considered).

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 15 Aug 2018 12:32 UTC**
The first problem with constant B.n in class C is that the new lookup rules are not applicable to class C but only to instances of class C.
A component "C c;" should be added to the example to fix this problem:

```
model A
  model B
    constant Integer n=1;
  equation
    assert(n==B.n, "n is not equal to B.n");
  end B;
end A;

class C = A.B(n=3);

C c;
```


The second problem is that it is not clear from lookup rules if the reference B.n is valid in the instance scope of c. The definition of B is found in a lexical scope that has no corresponding node in the instance tree of c. The specification does not say how this case should be interpreted. A cannot see a good interpretation:
 - not found, continue in the enclosing scope - does not work because the enclosing lexical scope may contain a wrong definition;
 - use the definition found in the lexical scope - not good because it would ignore modifications as in example below:

```
model A
  model B
    constant Integer n=1;
  equation
    assert(n==B.n, "n is not equal to B.n");
  end B;
end A;

model A1 = A(B(n=5));

model C = A1.B(n=3);

C c;
```

Here the lexical scope on n is A.B but it would be natural to use the value defined in A1.B.


The third problem is that Modelica provides multiple ways to reference same element:
 - using simple name (as n in the example)
 - using qualified name with first element found in the enclosing scope (as B.n in the example)
 - using global reference (.A.B.n)
 - using import
It is not clear that these references are not equivalent. Moreover, there was opinion among members of Modelica Association that they should be equivalent (but this was long time ago). It would be good to add some explanation to specification.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 15 Aug 2018 13:47 UTC**
Replying to [comment:6 pavel.grozman@…]:
> Sorry, I do not understand what is the correct flattening of m according to your explanation.
> 
> Variant 1:
> 
> ```
> Real m.x.x;
> Real m.x.y;
> ```
> 
> Variant 2:
> ```
> Real[3] m.x
> ```
> 
> My opinion is that the lookup in the instance tree (as defined in 5.3.1) gives variant 1, but the old lookup in the class tree gives variant 2.
> 
> I don't like the variant 1 because it does not correspond to what is defined (and checked) in package B.

Variant 2 is the correct one. The instance tree is not defined in 5.3.1, nor is the instance scope. In 5.3.1 there is a link to section 5.6.1.1 which defines The class tree. That link should be to 5.6.1.2 that defines both the instance tree and the scope instance (as given in my previous comment). If you follow those definitions you would get the Variant 2. Once again, you seem to interpret instance scope as the scope in which the whole instance is defined. The instance scope is **not** where the whole instance is defined. It is a node in the tree constructed for that instance and that corresponds to that particular element, so you should be looking in the scope where that particular element of the instance is defined.

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 15 Aug 2018 14:06 UTC**
**Disclaimer**

I can see where your concerns and the confusion is coming from. And let me say first, that I am myself very much in dislike of the type and instance separation/mess Modelica got into and would very much prefer a prototype-based language design instead of the crazy object-oriented-type mess we ended up with (particularly since at the end, Modelica has no "runtime" polymorphism for which reason the superimposition of a type system restricting instanziation is superfluous); but that is what we have and to change to prototypes now is a radical, not easy change.

**Why the original example is well-defined**

And now to the meat...
>The second problem is that it is not clear from lookup rules if the reference B.n is valid in the instance scope of c. The specification does not say how this case should be interpreted [...] [One interpretation might be:] use the definition found in the lexical scope - not good because it would ignore modifications [...]

To ignore such modifications is exactly what the standard explicitly says in 5.3.2, third case (which applies here):

>If the identifier denotes a class, that class is **temporarily flattened** (as if instantiating a component **without modifiers** of this class, see Section 7.2.2) and using the enclosing classes of the denoted class. The rest of the name (e.g., B or B.C] is looked up among the declared named elements of the temporary flattened class. If the class does not satisfy the requirements for a package, the lookup is restricted to encapsulated elements only. The class we look inside may not be partial in a simulation model.

Thus, it's not undefined but rather you don't like that modifications don't apply.

You can pack a value-reference in arbitrary many layers of nested models and packages with respective modifications; at the end all outside-its-lexical-scope modifications are ignored when that very reference starts with a type-name in its lexical scope (or in other words, only the modifications part of a value-reference's lexical scope are considered when that very reference starts with a type name according to its lexical scope). When referencing something for a value you must choose between either: (1) start with a component name in your lexical scope -> follows instance tree for resolution incorporating modifications, or, (2) start with a type name in your lexical scope -> follows ad-hoc flattened model of that very type for resolution, ignoring later modifications of it that are not within lexical scope.

**Bla bla**

There are good reasons for 5.3.2; in my opinion it's mostly because when we start permit mixing types and instances in non-simple names hell will freeze if only half-done. And that is what considering modifications in non-simple names containing types for value-references is all about: the modifications of these instances (queried for their value) in some non-instance context. That is something beyond sanity to define.

But if we go all-in and drop the distinction between types and instances (which is making the specification so hard in the first place), we have a very different language. Anyway, we are continuously pushing the boundaries in that direction (to enable flexible, easy modifications to existing models), and we will continuously be forced to back up because of Modelica's distinction of types and instances (**that is my personal opinion**).

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#46;grozman on 15 Aug 2018 15:01 UTC**
Replying to [comment:12 Christoff Bürger]:

> To ignore such modifications is exactly what the standard explicitly says in 5.3.2, third case (which applies here):
> 
> >If the identifier denotes a class, that class is **temporarily flattened** (as if instantiating a component **without modifiers** of this class, see Section 7.2.2) ...
> 
> Thus, it's not undefined but rather you don't like that modifications don't apply....

1. Before we can apply this rule to B.n, we should determine that B denotes a class. So we should find it using the lookup rule for simple names (5.3.1). My objection is that we will not find it in **instance scopes ... corresponding to lexically enclosing classes**. Do you mean that such an instance scope should be created "on the fly" for lexical scopes that have no corresponding node in the instance tree?

2. If we understand the rule "If the identifier denotes a class ..." literally, it should be applied even when it is found in the instance scope. Does this mean that the modifications of local classes are **always** ignored if the class is named by a composite names starting with a class (i.e., a composite name starting with a class is always equivalent to a global reference to the same class, even if it starts with a replaceable class?)

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#46;grozman on 16 Aug 2018 09:36 UTC**
Replying to [comment:11 Elena Shmoylova]:
> Replying to [comment:6 pavel.grozman@…]:
> > Sorry, I do not understand what is the correct flattening of m according to your explanation.
> > 
> > Variant 1:
> > 
> > {{{
> > Real m.x.x;
> > Real m.x.y;
> > }}}
> > 
> > Variant 2:
> > {{{
> > Real[3] m.x
> > }}}
> > 
> > My opinion is that the lookup in the instance tree (as defined in 5.3.1) gives variant 1, but the old lookup in the class tree gives variant 2.
> > 
> > I don't like the variant 1 because it does not correspond to what is defined (and checked) in package B.
> 
> Variant 2 is the correct one. The instance tree is not defined in 5.3.1, nor is the instance scope. In 5.3.1 there is a link to section 5.6.1.1 which defines The class tree. That link should be to 5.6.1.2 that defines both the instance tree and the scope instance (as given in my previous comment). If you follow those definitions you would get the Variant 2. Once again, you seem to interpret instance scope as the scope in which the whole instance is defined. The instance scope is **not** where the whole instance is defined. It is a node in the tree constructed for that instance and that corresponds to that particular element, so you should be looking in the scope where that particular element of the instance is defined.

I think that the problem is that the notion "instance scope" is not defined in the specification. Therefore I cannot understand your reasons.

I suggested that "instance scope" is a node in the instance tree. Is it true?

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 16 Aug 2018 10:00 UTC**
Replying to [comment:11 Elena Shmoylova]:
> Replying to [comment:6 pavel.grozman@…]:
> > Sorry, I do not understand what is the correct flattening of m according to your explanation.
> > 
> > Variant 1:
> > 
> > ```
> > Real m.x.x;
> > Real m.x.y;
> > }}}
> > 
> > Variant 2:
> > {{{
> > Real[3] m.x
> > }}}
> > 
> > My opinion is that the lookup in the instance tree (as defined in 5.3.1) gives variant 1, but the old lookup in the class tree gives variant 2.
> > 
> > I don't like the variant 1 because it does not correspond to what is defined (and checked) in package B.
> 
> Variant 2 is the correct one. The instance tree is not defined in 5.3.1, nor is the instance scope. In 5.3.1 there is a link to section 5.6.1.1 which defines The class tree. That link should be to 5.6.1.2 that defines both the instance tree and the scope instance (as given in my previous comment). If you follow those definitions you would get the Variant 2. Once again, you seem to interpret instance scope as the scope in which the whole instance is defined. The instance scope is **not** where the whole instance is defined. It is a node in the tree constructed for that instance and that corresponds to that particular element, so you should be looking in the scope where that particular element of the instance is defined.

I have tried to follow the definitions and got Variant 1 (assuming that the instance scope is a node in the instance tree). 
Here it the instance tree of m (as I understand it):

```
B.M m
   R x
     < to build the rest we should first determine what is R >
```

I am trying to find R for instantiation R x
 - First I have searched in the node corresponding the lexical scope of B.M. That is node m. No definition of R here;
 - Next I searched in the node corresponding to the enclosing scope of B.M, i.e., in the node corresponding to B. But the instance tree does not contain such node. So I skipped this step;
 - Next I have looked in the global scope and found the definition or R as a record.

What is my error?
```

**Comment by modelica-trac-importer** [comment]
**Comment by cbuerger on 16 Aug 2018 12:30 UTC**
I think the term instance scopes is troublesome and most likely not well-defined in the current standard. One maybe can avoid it altogether.

The rational for the term instance scopes is that 5.3.2 (including case 3) can be recursive. For example case 3, _temporary flatting as if instantiated without modifiers_, is followed by further lookup in the flattened model, which in turn may result in further applications of 5.3.2 and 5.3.1; thus, flattening happens on demand and may be nested. The idea of instance scopes for non-simple names that are composed of types is to handle such recursive on demand flattening; for non-simple names composed of components the idea is to consider the components' modifications. This does not mean that a non-simple name starting with a type cannot be modified; modifications according to the **lexical scope** it is used in are still considered. Big emphasis on modifications according to its **lexical scope**; non-lexical enclosed modifications due to instantiation in sibling-AST-branches are not relevant for non-simple names composed of types.

@Pavel: I have a lot of trouble to follow what you want when presenting your different possible interpretations. I think you are on a wrong track because you seem to start reasoning from an instance perspective. That is not the case; at its core Modelica is a lexical scoped language (putting `inner` and `outer` aside). Particularly, something referring to a type can never become a reference to a value and vice-versa, regardless of modifications, redeclarations or whatever. Likewise, a non-simple reference either is composed of a sequence of types or a sequence of components, never a mix thereof (please note, that constants are special in that a sequence of types can end referring to such); also this holds again regardless of modifications, redeclarations or whatever. Finally, a reference must refer to a type if its used for declaration and a value if its used within expressions (equations and algorithms); obviously, also this always holds regardless of further modifications, redeclarations or whatever. All this characteristics -- is something refering to a type or a value, is a type or value expected, is its name composed of types or components/instances -- can be decided by just looking at the lexical scope; reasoning about concrete instances is not required to decide any of these characteristics for a given reference in some context. At some point you either resolved the full name or figured out that it is invalid. Assuming it is valid, only thereafter -- after resolving it -- you start wondering what is the value of the thing I found (in case of expressions)/what is the class looking like (modifications for instantiation in case of declarations). Then the instance scopes kick in and you have to unravel instantiation. In that perspective you should try to interpret the standard and its lookup semantics.

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 16 Aug 2018 14:03 UTC**
Replying to [comment:16 Christoff Bürger]:
> @Pavel: I have a lot of trouble to follow what you want when presenting your different possible interpretations. I think you are on a wrong track because you seem to start reasoning from an instance perspective...

I did this only in order to understand what you mean by "lookup in the instance scope".

Actually my vision is that the lookup should be done only in the lexical scope, and only once for every occurrence of a name in the source code. As the result of lookup, the names are replaced with nodes.

When a class is copied (either by "extends" or by "short class definition" or when creating a component of that class), the meaning of the inherited names is derived by substituting the nodes that belong to the original class with the corresponding nodes in the derived class or component. No new lookup required at this stage.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 16 Aug 2018 14:14 UTC**
Replying to [comment:15 anonymous]:
> Replying to [comment:11 Elena Shmoylova]:
> > Replying to [comment:6 pavel.grozman@…]:
> > > Sorry, I do not understand what is the correct flattening of m according to your explanation.
> > > 
> > > Variant 1:
> > > 
> > > ```
> > > Real m.x.x;
> > > Real m.x.y;
> > > }}}
> > > 
> > > Variant 2:
> > > {{{
> > > Real[3] m.x
> > > }}}
> > > 
> > > My opinion is that the lookup in the instance tree (as defined in 5.3.1) gives variant 1, but the old lookup in the class tree gives variant 2.
> > > 
> > > I don't like the variant 1 because it does not correspond to what is defined (and checked) in package B.
> > 
> > Variant 2 is the correct one. The instance tree is not defined in 5.3.1, nor is the instance scope. In 5.3.1 there is a link to section 5.6.1.1 which defines The class tree. That link should be to 5.6.1.2 that defines both the instance tree and the scope instance (as given in my previous comment). If you follow those definitions you would get the Variant 2. Once again, you seem to interpret instance scope as the scope in which the whole instance is defined. The instance scope is **not** where the whole instance is defined. It is a node in the tree constructed for that instance and that corresponds to that particular element, so you should be looking in the scope where that particular element of the instance is defined.
> 
> I have tried to follow the definitions and got Variant 1 (assuming that the instance scope is a node in the instance tree). 
> Here it the instance tree of m (as I understand it):
> 
> {{{
> B.M m
>    R x
>      < to build the rest we should first determine what is R >
> ```
> 

Your instance tree is missing the property 2 (see [comment:5 comment 5]): "Each instance knows its source class definition from the class tree and its modification environment." The problem also is that property is not worded properly as I already mentioned in [comment:5 comment 5]. I think it should be each element in the instance tree or each node. At least, this is what was meant when the rewrite was discussed. It is this property that allows to search for R in the context where "R x" was defined. So the instance tree should be:
```
m < defined wherever R and B are defined
   R x < defined in B.M   
```
When you first look up R, you search in B.M because this is the node you are looking at, this is the instance scope. You cannot find it in B.M you continue searching in B.

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#x2e;grozman on 16 Aug 2018 14:30 UTC**
Replying to [comment:18 Elena Shmoylova]:
> When you first look up R, you search in B.M because this is the node you are looking at, this is the instance scope. You cannot find it in B.M you continue searching in B.

Why we should search in B? The specification explains (in 5.6.1.2) that
  If the name is not found the lookup is continued in the instance scope corresponding to the lexically enclosing class.

So we should search in the instance scope corresponding to B, not in B itself.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 16 Aug 2018 15:12 UTC**
Replying to [comment:19 pavel.grozman@…]:
> Replying to [comment:18 Elena Shmoylova]:
> > When you first look up R, you search in B.M because this is the node you are looking at, this is the instance scope. You cannot find it in B.M you continue searching in B.
> 
> Why we should search in B? The specification explains (in 5.6.1.2) that
>   If the name is not found the lookup is continued in the instance scope corresponding to the lexically enclosing class.
> 
> So we should search in the instance scope corresponding to B, not in B itself.

Sure, to be precise, you should also construct an instance tree for B.M, which also requires you to construct an instance tree for B. The instance scope for "R x" should point as defined in the instance tree for B.M. B.M itself would be a node in the instance tree for B. Note that the instance tree for B.M that corresponds to the instance m is not the same as an instance tree for m. In this example their content is identical but their scope is different. In a more general example, if there were modifications or extends clauses the content might be different as well.

For example:
```
package P

  model M1
    T1 x;
  end M1;

  model M2
    extends M1;
    T2 y;
  end M2;

  model C
    M2 m;
  end C;
end P;
```

The instance tree for P.C.m would be:
```
C.m < defined in P.C
   
   T1 x < defined in P.M1~P.M2~C.m
   T2 y < defined in P.M2~C.m
```
The instance tree P.M1~P.M2~C.m is the instance tree of P.M1 as extended by P.M2 and instantiated by C.m. The extension and instantiation can have modifications or redeclarations that would make this particular instance tree different from the class tree for P.M1. The scope of modifications would be where they are defined. But the scope of where the tree  P.M1~P.M2~C.M is still the same scope as for P.M1. T1 would be looked up in that instance tree. The content of that tree is different from the content of P.M2~C.m or C.m.

**Comment by modelica-trac-importer** [comment]
**Comment by pavel&#46;grozman on 26 Aug 2018 13:28 UTC**
Replying to [comment:20 Elena Shmoylova]:
> Replying to [comment:19 pavel.grozman@…]:
> > Replying to [comment:18 Elena Shmoylova]:
> > > When you first look up R, you search in B.M because this is the node you are looking at, this is the instance scope. You cannot find it in B.M you continue searching in B.
> > 
> > Why we should search in B? The specification explains (in 5.6.1.2) that
> >   If the name is not found the lookup is continued in the instance scope corresponding to the lexically enclosing class.
> > 
> > So we should search in the instance scope corresponding to B, not in B itself.
> 
> Sure, to be precise, you should also construct an instance tree for B.M,  ...

You can interpret the lookup in this way because you know its secret (or mystic) meaning from Modelica Oral Low.

Unfortunately, I have only access to the written specification and have to implement Modelica translator based on this specification.

I understand that it is not easy to explain the meaning of "Lookup in the instance scope". Maybe because no such scope exists in the simple sense of this word (as a portion of Modelica source code).

I would propose other definition of lookup (for simple names and the first identifier in composite names):

The name lookup associates the names in Modelica source code with nodes in class tree.

Then a node is created from a long class definition, the lookup for names used in the lexical scope of this definition is performed as follows
- Among declared named elements (class-definition and component-declaration) of the class (including elements inherited from base-classes).
- Among the import names of qualified import statements in the _lexical_ scope. The import name of import A.B.C; is C and the import name of import D=A.B.C; is D.
- Among the public members of packages imported via unqualified import-statements in the _lexical_ scope. It is an error if this step produces matches from several unqualified imports.
- Unless the class is encapsulated, repeat the steps above in the lexical scope of the enclosing class definition
- Among the elements defined in the global scope

Then a node (with its sub-tree) is created from another node (via short class definition, extends or component declaration), the inherited association of names is revised by replacing nodes that belong to the inherited tree with the corresponding nodes in the new tree.

I hope that my definition is equivalent (or will be equivalent after some tuning) with the secret meaning of "lookup in the instance scope".

**Comment by modelica-trac-importer** [comment]
**Modified by eshmoylova on 28 Aug 2018 13:51 UTC**

**Comment by HansOlsson** [comment]
Language Design Group thinks this is a misunderstanding, but it could be unclear wording => will revisit.

**Comment by HansOlsson** [comment]
Language group:
Unclear wording has not been found yet, assert should fail.
@daiwen investigates.

**Comment by Daiwen** [comment]
I think we could do a better job at specifying what we mean by "the instance of the lexically enclosing class of the scope"  in section 5.6.2:

> [To resolve the names, a name lookup using the instance tree is performed, starting at the instance scope(unless the name is fully qualified) of the modification, algorithm or equation.  **If it is not found locally the search is continued at the instance of the lexically enclosing class of the scope** [this is normally not equal to the parent of the current instance], and then continued with their parents as described in section 5.3.If  the  found  component  is  an  outer  declaration,  the  search  is  continued  using  the  direct  parents  in  the instance tree (see section 5.4).  If the lookup has to look into a class which is not instantiated yet [or only partially instantiated], it is instantiated in place.]

We do give a hint that it is not always equal to the parent of the current instance, but in some cases the continuation happens in the class tree, I would argue that it is the case for the original example.
That is, in "the instance of the lexically enclosing class of the scope", the "scope" refers to the instance of `C`, the "lexically enclosing class of the scope" is `A` (because `C` is defined as `A.B` ....), and its "instance", in this example, really is the class definition in the class tree.
I am not the original poster, so I can't tell for sure that the problem comes from this section of the specs being overlooked or misunderstood, but OP's intuition seems to be that the specs say that the name lookups should be performed in the instance tree only. I think this paragraph hints at the fact that it is not always the case, but I think the wording is confusing.

Not also that the section is supposed to talk about the generation of the flat equation system, but it gives important information about the instantiation process as shown by OP's second example.
